README.md
# flask-kadabra
Flask extension for the kadabra metrics framework.


CHANGES
Kadabra Changelog
=================

Here you can see the full list of changes between each Flask-Kadabra release.

Version 0.1.0
-------------

First public preview release.


test_flask_kadabra.py
from flask import (Flask, g, Response, current_app)

from flask_kadabra import Kadabra, record_metrics
import kadabra

from mock import mock, MagicMock, call

import datetime

NOW = datetime.datetime.utcnow()

def get_app():
    return Flask(__name__)

@mock.patch('flask_kadabra.Kadabra.init_app')
def test_ctor_no_app(mock_init_app):
    kadabra = Kadabra()

    kadabra.init_app.assert_has_calls([])
    assert kadabra.app == None

@mock.patch('flask_kadabra.Kadabra.init_app')
def test_ctor_app(mock_init_app):
    app = MagicMock()
    kadabra = Kadabra(app)

    kadabra.init_app.assert_called_with(app, None)
    assert kadabra.app == app

@mock.patch('flask_kadabra.Kadabra.init_app')
def test_ctor_app_config(mock_init_app):
    app = MagicMock()
    config = MagicMock()
    kadabra = Kadabra(app, config)

    kadabra.init_app.assert_called_with(app, config)
    assert kadabra.app == app

@mock.patch('kadabra.Kadabra')
def test_init(mock_client):
    client = mock_client.return_value
    app = get_app()

    kadabra = Kadabra()
    kadabra.init_app(app)

    assert kadabra.app == app
    assert app.kadabra == client
    assert len(app.before_request_funcs[None]) == 1
    assert app.before_request_funcs[None][0].__name__ == 'initialize_metrics'
    assert len(app.after_request_funcs[None]) == 1
    assert app.after_request_funcs[None][0].__name__ == 'transport_metrics'

@mock.patch('kadabra.Kadabra')
def test_init_metrics(mock_client):
    client = mock_client.return_value
    client.metrics = MagicMock()
    client.send = MagicMock()

    app = get_app()
    @app.route('/')
    def test_route():
        return 'test'

    kadabra = Kadabra()
    kadabra.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.metrics.assert_called_with()
        client.send.assert_has_calls([])

@mock.patch('flask_kadabra._get_now', return_value=NOW)
@mock.patch('kadabra.Kadabra')
def test_transport_200(mock_client, mock_get_now):
    client = mock_client.return_value
    client.metrics = MagicMock()
    metrics = client.metrics.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()

    @app.route('/')
    @record_metrics
    def test_route():
        return 'test'

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.metrics.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)

@mock.patch('flask_kadabra._get_now', return_value=NOW)
@mock.patch('kadabra.Kadabra')
def test_transport_500(mock_client, mock_get_now):
    client = mock_client.return_value
    client.metrics = MagicMock()
    metrics = client.metrics.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()
    
    @app.route('/')
    @record_metrics
    def test_route():
        response = Response()
        response.status_code = 500
        return response

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.metrics.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 1),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)

@mock.patch('flask_kadabra._get_now', return_value=NOW)
@mock.patch('kadabra.Kadabra')
def test_transport_400(mock_client, mock_get_now):
    client = mock_client.return_value
    client.metrics = MagicMock()
    metrics = client.metrics.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()

    @app.route('/')
    @record_metrics
    def test_route():
        response = Response()
        response.status_code = 400
        return response

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.metrics.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 1)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)

@mock.patch('flask_kadabra._get_now', return_value=NOW)
@mock.patch('kadabra.Kadabra')
def test_transport_disable(mock_client, mock_get_now):
    client = mock_client.return_value
    client.metrics = MagicMock()
    metrics = client.metrics.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()
    app.config["DISABLE_KADABRA"] = True

    @app.route('/')
    @record_metrics
    def test_route():
        return 'test'

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.metrics.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_has_calls([])


tests/test_kadabra.py
from flask import (Flask, g, Response, request, current_app)

from flask_kadabra import Kadabra, record_metrics
import kadabra

from mock import mock, MagicMock, call

import datetime

NOW = datetime.datetime.utcnow()

class MockDatetime(datetime.datetime):
    "A fake replacement for date that can be mocked for testing."
    def __new__(cls, *args, **kwargs):
        return datetime.datetime.__new__(datetime.datetime, *args, **kwargs)

    @classmethod
    def utcnow(cls):
        return NOW

def get_app():
    return Flask(__name__)

@mock.patch('flask_kadabra.Kadabra.init_app')
def test_ctor_no_app(mock_init_app):
    kadabra = Kadabra()

    kadabra.init_app.assert_has_calls([])
    assert kadabra.app == None

@mock.patch('flask_kadabra.Kadabra.init_app')
def test_ctor_no_app(mock_init_app):
    app = MagicMock()
    kadabra = Kadabra(app)

    kadabra.init_app.assert_called_with(app)
    assert kadabra.app == app

@mock.patch('kadabra.Client')
def test_init(mock_client):
    client = mock_client.return_value
    app = get_app()

    kadabra = Kadabra()
    kadabra.init_app(app)

    assert kadabra.app == app
    assert app.kadabra == client
    assert len(app.before_request_funcs[None]) == 1
    assert app.before_request_funcs[None][0].__name__ == 'initialize_metrics'
    assert len(app.after_request_funcs[None]) == 1
    assert app.after_request_funcs[None][0].__name__ == 'transport_metrics'

@mock.patch('kadabra.Client')
def test_init_metrics(mock_client):
    client = mock_client.return_value
    client.get_collector = MagicMock()
    client.send = MagicMock()

    app = get_app()
    @app.route('/')
    def test_route():
        return 'test'

    kadabra = Kadabra()
    kadabra.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.get_collector.assert_called_with()
        client.send.assert_has_calls([])

@mock.patch('flask_kadabra.datetime.datetime', MockDatetime)
@mock.patch('kadabra.Client')
def test_transport_200(mock_client):
    client = mock_client.return_value
    client.get_collector = MagicMock()
    metrics = client.get_collector.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()

    @app.route('/')
    @record_metrics()
    def test_route():
        return 'test'

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.get_collector.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)

@mock.patch('flask_kadabra.datetime.datetime', MockDatetime)
@mock.patch('kadabra.Client')
def test_transport_500(mock_client):
    client = mock_client.return_value
    client.get_collector = MagicMock()
    metrics = client.get_collector.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()
    
    @app.route('/')
    @record_metrics()
    def test_route():
        response = Response()
        response.status_code = 500
        return response

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.get_collector.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 1),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)

@mock.patch('flask_kadabra.datetime.datetime', MockDatetime)
@mock.patch('kadabra.Client')
def test_transport_400(mock_client):
    client = mock_client.return_value
    client.get_collector = MagicMock()
    metrics = client.get_collector.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()

    @app.route('/')
    @record_metrics()
    def test_route():
        response = Response()
        response.status_code = 400
        return response

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.get_collector.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 1)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)

@mock.patch('flask_kadabra.datetime.datetime', MockDatetime)
@mock.patch('kadabra.Client')
def test_transport_disable(mock_client):
    client = mock_client.return_value
    client.get_collector = MagicMock()
    metrics = client.get_collector.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    app = get_app()
    app.config["DISABLE_KADABRA"] = True

    @app.route('/')
    @record_metrics()
    def test_route():
        return 'test'

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.get_collector.assert_called_with()
        metrics.set_dimension.assert_called_with("method", "test_route")
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_has_calls([])

@mock.patch('flask_kadabra.datetime.datetime', MockDatetime)
@mock.patch('kadabra.Client')
def test_decorator_dimensions(mock_client):
    client = mock_client.return_value
    client.get_collector = MagicMock()
    metrics = client.get_collector.return_value
    metrics.add_count = MagicMock()
    metrics.set_timer = MagicMock()
    metrics.set_dimension = MagicMock()
    metrics.close = MagicMock()
    closed = metrics.close.return_value
    client.send = MagicMock()

    testDimensionName = "testDimensionName"
    testDimensionValue = "testDimensionValue"

    app = get_app()

    @app.route('/')
    @record_metrics({testDimensionName: testDimensionValue})
    def test_route():
        return 'test'

    unit = Kadabra()
    unit.init_app(app)

    with app.test_client() as c:
        c.get('/')
        client.get_collector.assert_called_with()
        metrics.set_dimension.assert_has_calls([
                call("method", "test_route"),
                call(testDimensionName, testDimensionValue)])
        metrics.set_timer.assert_called_with("RequestTime", NOW - NOW,
                kadabra.Units.MILLISECONDS)
        metrics.add_count.assert_has_calls([
                call("Failure", 0),
                call("ClientError", 0)])
        metrics.close.assert_called_with()
        client.send.assert_called_with(closed)


flask_kadabra.py
import datetime
import kadabra

from flask import g, current_app
from flask import _app_ctx_stack as stack

from functools import wraps

__version__ = '0.1.0'

class Kadabra(object):
    """This object holds ties the Flask application object to the Kadabra
    library. Each app object gets its own :class:`~kadabra.Kadabra` instance,
    which it uses to generate a :class:`~kadabra.client.MetricsCollector` for
    each request.

    :param app: The Flask application object to initialize.
    :type app: ~flask.Flask

    :param config: Dictionary of configuration to use for the
                   :class:`~kadabra.Kadabra` client API. For information on
                   the acceptable values see :ref:`kadabra:configuration`.
    :type config: dict
    """
    def __init__(self, app=None, config=None):
        self.app = app
        if app is not None:
            self.init_app(app, config)

    def init_app(self, app, config=None):
        """Configure the application to use Kadabra. Requests will have access
        to a :class:`~kadabra.client.MetricsCollector` via the ``metrics``
        attribute of Flask's :data:`~flask.g` object. You can record
        metrics anywhere in the context of a request like so::

            ...
            g.metrics.add_count("userSignup", 1)
            ...

        The metrics object will be closed and sent at the end of the
        request if any view that handles the request has been annotated with
        :data:`~flask_kadabra.record_metrics`."""
        app.kadabra = kadabra.Kadabra(config)
        self.app = app

        @app.before_request
        def initialize_metrics():
            ctx = stack.top
            if ctx is not None:
                ctx.kadabra_request_start_time = _get_now()
                g.metrics = current_app.kadabra.metrics()

        @app.after_request
        def transport_metrics(response):
            # Only send the metrics if the current view has "opted in".
            ctx = stack.top
            if ctx is not None and getattr(ctx, "enable_kadabra", False):
                end_time = _get_now()
                g.metrics.set_timer("RequestTime",
                        (end_time - ctx.kadabra_request_start_time),
                        kadabra.Units.MILLISECONDS)

                failure = 0
                client_error = 0
                if response.status_code >= 500:
                    failure = 1
                elif response.status_code >= 400:
                    client_error = 1

                g.metrics.add_count("Failure", failure)
                g.metrics.add_count("ClientError", client_error)

                closed = g.metrics.close()
                if not current_app.config.get("DISABLE_KADABRA"):
                    current_app.kadabra.send(closed)
            return response

def record_metrics(func):
    """Views that are annotated with this decorator will cause any request they
    handle to send all metrics collected via the Kadabra client API. For
    example::

        @api.route('/')
        @record_metrics
        def index():
            return 'Hello, world!'

    :param func: The view function to decorate.
    :type func: function
    """
    @wraps(func)
    def decorated_view(*args, **kwargs):
        ctx = stack.top
        if ctx is not None:
            ctx.enable_kadabra = True
            g.metrics.set_dimension("method", func.__name__)
        return func(*args, **kwargs)
    return decorated_view

def _get_now():
    return datetime.datetime.utcnow()
